### PR TITLE
feat: Enable bazel generation

### PR DESCRIPTION
**Pull request by mrmeku**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [ ] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`. If contributing template-only or documentation-only changes which will change sample output, be sure to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) first.
- [x] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.1.x`, `5.0.x`. Default: `master`.
- [ ] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

With this PR you are now able to use bazel to run the openapi generatator at build time

Example usage can be found in `modules/openapi-generator-bazel/src/test/BUILD`

```
openapi_generator(
    name = "petstore",
    language = "go",
    spec = "petstore.yaml",
)
```

### CODE REVIEW

File: modules/openapi-generator-bazel/src/defs.bzl [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        inputs = inputs + [declared_dir],
+    )
+
+    return DefaultInfo(files = depset([
+        declared_dir,
+        # TODO: Split off so that non consumers of srcjars do not create them
+        ctx.outputs.codegen,
+    ]))
+
+# taken from rules_scala
```
[the remark this answers is not in the document]
Comment (mrmeku) [reply modules/openapi-generator-bazel/src/defs.bzl]: Doesn't exist anymore it seems

**Comment by mrmeku** [comment]
@wing328

File: modules/openapi-generator-bazel/src/defs.bzl [hunk 2]
```diff
@@ -0,0 +1,168 @@
+def openapi_generator_repositories(swagger_openapi_generator_cli_version = "4.0.3", swagger_openapi_generator_cli_sha1 = "4a7e6d7c82df64a1d869e68f28f23e6afd0f9d85", prefix = "io_bazel_rules_openapi_generator"):
```
Comment (jimschubert) [inline modules/openapi-generator-bazel/src/defs.bzl]: Should remove swagger references. Although OpenAPI 2.0 and Swagger 2.0 are one and the same, Swagger itself is a trademark of SmartBear and is in no relation to this project.

File: modules/openapi-generator-bazel/src/defs.bzl [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    if ctx.attr.invoker_package:
+        gen_cmd += " --invoker-package {package}".format(
+            package = ctx.attr.invoker_package,
+        )
+    if ctx.attr.model_package:
+        gen_cmd += " --model-package {package}".format(
+            package = ctx.attr.model_package,
+        )
+
+    # fixme: by default, swagger-codegen is rather verbose. this helps with that but can also mask useful error messages
```
Comment (jimschubert) [inline modules/openapi-generator-bazel/src/defs.bzl]: Change `swagger-codegen`

File: modules/openapi-generator-bazel/src/defs.bzl [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        # openapi spec file
+        "spec": attr.label(
+            mandatory = True,
+            allow_single_file = [
+                ".json",
+                ".yaml",
+            ],
+        ),
+        # language to generate
+        "language": attr.string(mandatory = True),
```
Comment (jimschubert) [inline modules/openapi-generator-bazel/src/defs.bzl]: This option has been deprecated for a while, and removed in more recent versions. l/language has been replaced with g/generator-name in CLI options.

**Comment by jimschubert** [comment]
Thanks for the contribution! This looks great. I've wanted to look into Bazel for a while, and I guess that time is now. ;)

File: modules/openapi-generator-bazel/src/defs.bzl (hunk 2 above)
Comment (mrmeku) [reply modules/openapi-generator-bazel/src/defs.bzl]: Will do, thanks for the catch

File: modules/openapi-generator-bazel/src/defs.bzl (hunk 4 above)
Comment (mrmeku) [reply modules/openapi-generator-bazel/src/defs.bzl]: Done! I was actually using the -g flag in my invocation, I just didn't reflect that in the bazel rule :p

**Comment by mrmeku** [comment]
> Thanks for the contribution! This looks great. I've wanted to look into Bazel for a while, and I guess that time is now. ;)

If you ever want to talk bazel and or hack bazel you know where to find me

**Comment by wing328** [comment]
cc @OpenAPITools/generator-core-team

**Comment by wing328** [comment]
@mrmeku when you've time, can you please PM me via Slack: https://join.slack.com/t/openapi-generator/shared_invite/enQtNzAyNDMyOTU0OTE1LTY5ZDBiNDI5NzI5ZjQ1Y2E5OWVjMjZkYzY1ZGM2MWQ4YWFjMzcyNDY5MGI4NjQxNDBiMTlmZTc5NjY2ZTQ5MGM?

I've a few simple questions for you if you don't mind.

**Comment by wing328** [comment]
FYI. I've added you as the template creator of asciidoc in https://github.com/OpenAPITools/openapi-generator/commit/8f433739481295b420c2cffc09fbbb6e3676ae8d

**Comment by wing328** [comment]
@mrmeku I think all "language" needs to be replaced with "generator" in the BUILD file

Is there any bazel command to list all the available"attribute"?

**Comment by jimschubert** [comment]
Also, I've spent a little time trying to evaluate and couldn't even build Bazel examples locally. Since I'm not familiar with Bazel, I'm not sure how to evaluate.

I'm wondering if you have a minimal example project where this can be evaluated, with some example command usage?

**Comment by mrmeku** [comment]
@jimschubert Thanks for reviewing, sorry that you had issue in getting the bazel code to build locally. What were you trying? What can I provide to you that would be helpful? Also, would you like to move the conversation to slack for more immediate responses? (I'm on the slack channel for this project)

Cheers!

**Comment by mrmeku** [comment]
@jimschubert and anyone else reviewing, to build the test cases install bazel for your system and run
`bazel build modules/openapi-generator-bazel/src/test/...`

**Comment by mrmeku** [comment]
Any word on merging this PR?

**Comment by macjohnny** [comment]
@wing328 @jimschubert can we merge this and see how it goes?

**Comment by mrmeku** [comment]
Yo, @wing328 @jimschubert can we get an update?

**Comment by wing328** [comment]
@mrmeku thanks for pinging me but I'm no expert in bazel.

@jimschubert can you please review when you've time?

**Comment by jimschubert** [comment]
@mrmeku Thank you for providing a test case. I'm still very confused, because even after running the sample command you've provided, I see output:

```
➜  openapi-generator git:(mrmeku-bazel) bazel build modules/openapi-generator-bazel/src/test/...
Extracting Bazel installation...
Starting local Bazel server and connecting to it...
INFO: Analyzed 5 targets (17 packages loaded, 466 targets configured).
INFO: Found 5 targets...
INFO: Elapsed time: 153.316s, Critical Path: 11.96s
INFO: 5 processes: 5 darwin-sandbox.
INFO: Build completed successfully, 6 total actions
```

But there are no generated files that I can find.  Given this and the last time I reviewed when I attempted to build a project I found on GitHub and it similarly did nothing, I think there may be something wrong with my installation.

I do have some concerns that the above build took 3 minutes to execute. Our CI is already overloaded, and adding this as part of the build would overload it more.  I'm wondering if it might make sense to add this as a separate repository under the OpenAPI Tools org? I feel like that would allow us to trigger builds on the repository using GitHub actions, and separate concerns a little. I think it would also allow us to define an owner or team with write access to that repository to limit Issue/PRs. What are your thoughts on that?

@OpenAPITools/generator-core-team I'm wondering if anyone else on the core team might be familiar with Bazel, or able to run the above test commands and generate outputs?

**Comment by jimschubert** [comment]
Oh. I've switched a branch and found that I _do_ have artifacts from this build. It's odd because I ran a `find` command in my projects directory, which is one level up from the openapi-generator directory, and I didn't find any of the outputs defined in the test file.

I do see:

```
cd bazel-genfiles/modules/openapi-generator-bazel/src/test/petstore
➜  petstore cd ..
➜  test ls -la *
petstore:
total 136
dr-xr-xr-x  18 jim  wheel    576 Oct 17 21:20 .
drwxr-xr-x   7 jim  wheel    224 Oct 17 21:20 ..
-r-xr-xr-x   1 jim  wheel    266 Oct 17 21:20 .gitignore
dr-xr-xr-x   3 jim  wheel     96 Oct 17 21:20 .openapi-generator
-r-xr-xr-x   1 jim  wheel   1040 Oct 17 21:20 .openapi-generator-ignore
-r-xr-xr-x   1 jim  wheel     71 Oct 17 21:20 .travis.yml
-r-xr-xr-x   1 jim  wheel   1449 Oct 17 21:20 README.md
dr-xr-xr-x   3 jim  wheel     96 Oct 17 21:20 api
-r-xr-xr-x   1 jim  wheel   9582 Oct 17 21:20 api_pets.go
-r-xr-xr-x   1 jim  wheel  12202 Oct 17 21:20 client.go
-r-xr-xr-x   1 jim  wheel   2126 Oct 17 21:20 configuration.go
dr-xr-xr-x   5 jim  wheel    160 Oct 17 21:20 docs
-r-xr-xr-x   1 jim  wheel   1662 Oct 17 21:20 git_push.sh
-r-xr-xr-x   1 jim  wheel    181 Oct 17 21:20 go.mod
-r-xr-xr-x   1 jim  wheel   1095 Oct 17 21:20 go.sum
-r-xr-xr-x   1 jim  wheel    332 Oct 17 21:20 model_error.go
-r-xr-xr-x   1 jim  wheel    355 Oct 17 21:20 model_pet.go
-r-xr-xr-x   1 jim  wheel   1262 Oct 17 21:20 response.go

petstore_java:
total 120
dr-xr-xr-x  18 jim  wheel    576 Oct 17 21:20 .
drwxr-xr-x   7 jim  wheel    224 Oct 17 21:20 ..
-r-xr-xr-x   1 jim  wheel    291 Oct 17 21:20 .gitignore
dr-xr-xr-x   3 jim  wheel     96 Oct 17 21:20 .openapi-generator
-r-xr-xr-x   1 jim  wheel   1040 Oct 17 21:20 .openapi-generator-ignore
-r-xr-xr-x   1 jim  wheel    411 Oct 17 21:20 .travis.yml
-r-xr-xr-x   1 jim  wheel   3127 Oct 17 21:20 README.md
-r-xr-xr-x   1 jim  wheel   3179 Oct 17 21:20 build.gradle
-r-xr-xr-x   1 jim  wheel    969 Oct 17 21:20 build.sbt
dr-xr-xr-x   5 jim  wheel    160 Oct 17 21:20 docs
-r-xr-xr-x   1 jim  wheel   1662 Oct 17 21:20 git_push.sh
dr-xr-xr-x   3 jim  wheel     96 Oct 17 21:20 gradle
-r-xr-xr-x   1 jim  wheel     50 Oct 17 21:20 gradle.properties
-r-xr-xr-x   1 jim  wheel   5305 Oct 17 21:20 gradlew
-r-xr-xr-x   1 jim  wheel   2176 Oct 17 21:20 gradlew.bat
-r-xr-xr-x   1 jim  wheel  10380 Oct 17 21:20 pom.xml
-r-xr-xr-x   1 jim  wheel     40 Oct 17 21:20 settings.gradle
dr-xr-xr-x   4 jim  wheel    128 Oct 17 21:20 src

petstore_java_bigdecimal:
total 120
dr-xr-xr-x  18 jim  wheel    576 Oct 17 21:20 .
drwxr-xr-x   7 jim  wheel    224 Oct 17 21:20 ..
-r-xr-xr-x   1 jim  wheel    291 Oct 17 21:20 .gitignore
dr-xr-xr-x   3 jim  wheel     96 Oct 17 21:20 .openapi-generator
-r-xr-xr-x   1 jim  wheel   1040 Oct 17 21:20 .openapi-generator-ignore
-r-xr-xr-x   1 jim  wheel    411 Oct 17 21:20 .travis.yml
-r-xr-xr-x   1 jim  wheel   3127 Oct 17 21:20 README.md
-r-xr-xr-x   1 jim  wheel   3179 Oct 17 21:20 build.gradle
-r-xr-xr-x   1 jim  wheel    969 Oct 17 21:20 build.sbt
dr-xr-xr-x   5 jim  wheel    160 Oct 17 21:20 docs
-r-xr-xr-x   1 jim  wheel   1662 Oct 17 21:20 git_push.sh
dr-xr-xr-x   3 jim  wheel     96 Oct 17 21:20 gradle
-r-xr-xr-x   1 jim  wheel     50 Oct 17 21:20 gradle.properties
-r-xr-xr-x   1 jim  wheel   5305 Oct 17 21:20 gradlew
-r-xr-xr-x   1 jim  wheel   2176 Oct 17 21:20 gradlew.bat
-r-xr-xr-x   1 jim  wheel  10380 Oct 17 21:20 pom.xml
-r-xr-xr-x   1 jim  wheel     40 Oct 17 21:20 settings.gradle
dr-xr-xr-x   4 jim  wheel    128 Oct 17 21:20 src

petstore_java_feign:
total 120
dr-xr-xr-x  17 jim  wheel    544 Oct 17 21:20 .
drwxr-xr-x   7 jim  wheel    224 Oct 17 21:20 ..
-r-xr-xr-x   1 jim  wheel    291 Oct 17 21:20 .gitignore
dr-xr-xr-x   3 jim  wheel     96 Oct 17 21:20 .openapi-generator
-r-xr-xr-x   1 jim  wheel   1040 Oct 17 21:20 .openapi-generator-ignore
-r-xr-xr-x   1 jim  wheel    411 Oct 17 21:20 .travis.yml
-r-xr-xr-x   1 jim  wheel   1000 Oct 17 21:20 README.md
-r-xr-xr-x   1 jim  wheel   3692 Oct 17 21:20 build.gradle
-r-xr-xr-x   1 jim  wheel   1414 Oct 17 21:20 build.sbt
-r-xr-xr-x   1 jim  wheel   1662 Oct 17 21:20 git_push.sh
dr-xr-xr-x   3 jim  wheel     96 Oct 17 21:20 gradle
-r-xr-xr-x   1 jim  wheel     50 Oct 17 21:20 gradle.properties
-r-xr-xr-x   1 jim  wheel   5305 Oct 17 21:20 gradlew
-r-xr-xr-x   1 jim  wheel   2176 Oct 17 21:20 gradlew.bat
-r-xr-xr-x   1 jim  wheel  11075 Oct 17 21:20 pom.xml
-r-xr-xr-x   1 jim  wheel     40 Oct 17 21:20 settings.gradle
dr-xr-xr-x   4 jim  wheel    128 Oct 17 21:20 src

petstore_java_no_tests:
total 120
dr-xr-xr-x  18 jim  wheel    576 Oct 17 21:20 .
drwxr-xr-x   7 jim  wheel    224 Oct 17 21:20 ..
-r-xr-xr-x   1 jim  wheel    291 Oct 17 21:20 .gitignore
dr-xr-xr-x   3 jim  wheel     96 Oct 17 21:20 .openapi-generator
-r-xr-xr-x   1 jim  wheel   1040 Oct 17 21:20 .openapi-generator-ignore
-r-xr-xr-x   1 jim  wheel    411 Oct 17 21:20 .travis.yml
-r-xr-xr-x   1 jim  wheel   3127 Oct 17 21:20 README.md
-r-xr-xr-x   1 jim  wheel   3179 Oct 17 21:20 build.gradle
-r-xr-xr-x   1 jim  wheel    969 Oct 17 21:20 build.sbt
dr-xr-xr-x   5 jim  wheel    160 Oct 17 21:20 docs
-r-xr-xr-x   1 jim  wheel   1662 Oct 17 21:20 git_push.sh
dr-xr-xr-x   3 jim  wheel     96 Oct 17 21:20 gradle
-r-xr-xr-x   1 jim  wheel     50 Oct 17 21:20 gradle.properties
-r-xr-xr-x   1 jim  wheel   5305 Oct 17 21:20 gradlew
-r-xr-xr-x   1 jim  wheel   2176 Oct 17 21:20 gradlew.bat
-r-xr-xr-x   1 jim  wheel  10380 Oct 17 21:20 pom.xml
-r-xr-xr-x   1 jim  wheel     40 Oct 17 21:20 settings.gradle
dr-xr-xr-x   4 jim  wheel    128 Oct 17 21:20 src
```

So, I guess the remaining question is whether or not we can extract the code to a separate repo under the org. @mrmeku if we were to move it to a separate repo, would you be willing to be a code owner for that repo?

**Comment by mrmeku** [comment]
Sure I don't see why not. If you provision me a repo I'll push the code there

**Comment by jimschubert** [comment]
@mrmeku I've invited you to a new group in the org called `bazel` and created the repo at https://github.com/OpenAPITools/openapi-generator-bazel. Membership in the group gives write permission to the repo.

**Pull request closed without merging**